appdrawdance/CMakeLists.txt

```diff
@@ -1,101 +1,61 @@
-    drawdance_lua/lua_util.h)
-
-add_clang_format_files("${drawdance_lua_sources}" "${drawdance_lua_headers}")
-
-add_library(drawdance_lua STATIC "${drawdance_lua_sources}"
-                                 "${drawdance_lua_headers}")
-set_dp_target_properties(drawdance_lua)
-target_include_directories(drawdance_lua PRIVATE "${CMAKE_BINARY_DIR}/config")
-target_include_directories(drawdance_lua INTERFACE drawdance_lua)
-target_link_libraries(drawdance_lua PUBLIC dpclient dpmsg lua)
-
-if(NOT DRAWDANCE_EMSCRIPTEN)
-    target_compile_definitions(drawdance_lua PRIVATE DRAWDANCE_IMGUI)
-    target_link_libraries(drawdance_lua PUBLIC drawdance_lua_cxx)
-endif()
-
-
-set(drawdance_sources
+dp_add_executable(drawdance)
+target_sources(drawdance PRIVATE
     drawdance/app.c
-    drawdance/canvas_renderer.c
-    drawdance/gl.c
-    drawdance/main.c
-    drawdance/ui.c)
-
-set(drawdance_headers
     drawdance/app.h
+    drawdance/canvas_renderer.c
     drawdance/canvas_renderer.h
     drawdance/emproxy.h
+    drawdance/gl.c
     drawdance/gl.h
-    drawdance/ui.h)
-
-add_clang_format_files("${drawdance_sources}" "${drawdance_headers}")
-
-add_executable(drawdance "${drawdance_sources}" "${drawdance_headers}")
-set_dp_target_properties(drawdance)
-target_link_libraries(drawdance PUBLIC drawdance_lua dpclient dpengine dpgles2)
-
-if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
-    target_compile_definitions(drawdance PRIVATE DP_GL_CHECKS)
-endif()
-
-if(DRAWDANCE_EMSCRIPTEN)
+    drawdance/main.c
+    drawdance/ui.c
+    drawdance/ui.h
+    drawdance_gui/gui.cpp
+    drawdance_gui/gui.h
+    drawdance_lua/client.c
+    drawdance_lua/document.c
+    drawdance_lua/imgui.cpp
+    drawdance_lua/imgui_gen.cpp
+    drawdance_lua/lua_bindings.c
+    drawdance_lua/lua_bindings.h
+    drawdance_lua/lua_util.h
+    drawdance_lua/message.c
+    drawdance_lua/ui.c
+)
+target_include_directories(drawdance PRIVATE
+    drawdance
+    drawdance_gui
+    drawdance_lua
+    "${CMAKE_BINARY_DIR}/config"
+)
+target_link_libraries(drawdance PRIVATE
+    dpcommon
+    dpclient
+    dpengine
+    dpmsg
+    lua
+    parson
+    imgui
+    dpgles2
+    SDL2::SDL2
+)
+target_compile_definitions(drawdance PRIVATE $<$<CONFIG:Debug>:DP_GL_CHECKS>)
+
+if(EMSCRIPTEN)
     target_link_options(drawdance PRIVATE
-        "SHELL:-s EXPORTED_FUNCTIONS=_main,_DP_send_from_browser"
-        "SHELL:-s EXPORTED_RUNTIME_METHODS=ALLOC_NORMAL,allocate,intArrayFromString"
-        "SHELL:--preload-file '${CMAKE_SOURCE_DIR}/appdrawdance/lua@appdrawdance/lua'")
+        "-s PROXY_TO_PTHREAD"
+        "-s ALLOW_BLOCKING_ON_MAIN_THREAD=0"
+        "-s ALLOW_MEMORY_GROWTH=1"
+        "-s MODULARIZE=1"
+        "-s EXPORT_NAME=createModule"
+        "-s DYNAMIC_EXECUTION=0"
+        "-s EXPORTED_FUNCTIONS=_main,_DP_send_from_browser"
+        "-s EXPORTED_RUNTIME_METHODS=ALLOC_NORMAL,allocate,intArrayFromString"
+        "--preload-file '${CMAKE_SOURCE_DIR}/appdrawdance/lua@appdrawdance/lua'"
```

Comment:
This args need to be written like this:
```cmake
        "-sPROXY_TO_PTHREAD"
        "-sALLOW_BLOCKING_ON_MAIN_THREAD=0"
        "-sALLOW_MEMORY_GROWTH=1"
        "-sMODULARIZE=1"
        "-sEXPORT_NAME=createModule"
        "-sDYNAMIC_EXECUTION=0"
        "-sEXPORTED_FUNCTIONS=_main,_DP_send_from_browser"
        "-sEXPORTED_RUNTIME_METHODS=ALLOC_NORMAL,allocate,intArrayFromString"
        "--preload-file=${CMAKE_SOURCE_DIR}/appdrawdance/lua@appdrawdance/lua"
```

Otherwise CMake will *helpfully* split and deduplicate them, so you end up with a single `-s` followed by a bunch of garbage orphaned arguments. That was the reason for the `SHELL:` stuff previously, but nowadays Emscripten supports passing the flags without spaces in-between, so we can just use those.

Reply:
Could you verify which CMake and emsdk versions you are using and which generator and which OS? I want to make sure I understand where the problem is coming from, since this command is executed without any funny stuff like that happening here:

(line endings added for clarity)

```sh
cmake -E cmake_link_script CMakeFiles/drawdance.dir/link.txt --verbose=1 \
  emsdk/upstream/emscripten/em++ -O2 -g -DNDEBUG
    "-s PROXY_TO_PTHREAD"
    "-s ALLOW_BLOCKING_ON_MAIN_THREAD=0"
    "-s ALLOW_MEMORY_GROWTH=1"
    "-s MODULARIZE=1"
    "-s EXPORT_NAME=createModule"
    "-s DYNAMIC_EXECUTION=0"
    "-s EXPORTED_FUNCTIONS=_main,_DP_send_from_browser"
    "-s EXPORTED_RUNTIME_METHODS=ALLOC_NORMAL,allocate,intArrayFromString"
    "--preload-file 'Drawdance/appdrawdance/lua@appdrawdance/lua'"
    -s FETCH
    -lwebsocket.js
    @CMakeFiles/drawdance.dir/objects1.rsp -o drawdance.js @CMakeFiles/drawdance.dir/linklibs.rsp
```

Reply:
I'm on CMake 3.26.1, emsdk 3.1.34 and Ninja as the generator on Arch Linux. Makefile generator shows the same behavior. But that's intentional and documented (and utterly confounding): <https://cmake.org/cmake/help/latest/command/add_link_options.html#option-de-duplication>

You can try it both ways and `grep createModule $builddir/appdrawdance/drawdance.js`. With the current version that uses spaces between the arguments, it will find nothing. Without spaces, the export will be there as requested. Sadly, since emcc wraps so many different tools that mutate and mangle the arguments along the way, it fails to warn about the arguments dropped on the floor, so you only see the result when running it and being told that `createModule` is undefined.

Reply:
I don’t understand how it would be doing that to quoted options, they are quoted which means they are one item, not mul… whatever, argh. Why is CMake. I have made this change.